File: 3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.cpp
class Solution {
public:
    long long find(string &s,int k){
        unordered_map<char,int>mp;
        string vowel="aeiou";
        int consonants=0;
        long long count=0;
        for(int i=0,j=0;i<s.size();i++){
            if(vowel.find(s[i])!=string::npos){
                mp[s[i]]++;
            }
            else{
                consonants++;
            }
            while(mp.size()==5&&consonants>k){
                if(vowel.find(s[j])!=string::npos){
                    if(--mp[s[j]]==0){
                        mp.erase(s[j]);
                    }
                }
                else{
                    consonants--;
                }
                j++;
            }
            count+=(i-j+1);
        }
        return count;
    }
    long long countOfSubstrings(string word, int k) {
        return find(word,k)-find(word,k-1);
    }
};

File: 3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.py
class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:

        vowel = {c:0 for c in "aeiou"}
        v = {c:0 for c in "aeiou"}
        c = 0
        cons = 0
        i = 0
        minwind = 0
        ans = 0
        for j, char in enumerate(word):
            if char in vowel:
                vowel[char] += 1
                v[char] += 1
            else:
                cons += 1
                c += 1

            while cons > k:
                if word[i] in vowel:
                    vowel[word[i]] -= 1
                    if i == minwind:
                        v[word[i]] -= 1
                else:
                    cons -= 1
                    if i == minwind:
                        c -= 1
                if i == minwind:
                    minwind += 1
                i += 1
                
            # print(ans, cons, vowel, char)
            if cons == k and all(val > 0 for val in vowel.values()):
                
                while c == k and all(val > 0 for val in v.values()):
                    if word[minwind] in vowel:
                        v[word[minwind]] -= 1
                    else:
                        c -= 1
                    minwind += 1
                
                ans += minwind - i

        return ans